File: server/src/types/shipment.ts
export interface Shipment {
  id?: number;
  containerId: string;
  originPort: string;
  destinationPort: string;
  departureDate: Date;
  status: ShipmentStatus;
  eta: Date;
  lastKnownLat?: number | null;
  lastKnownLng?: number | null;
  createdAt?: Date;
  updatedAt?: Date;
}

export type ShipmentStatus = 'In Transit' | 'Arrived at Port' | 'Delayed' | 'Delivered';

export interface CreateShipmentRequest {
  containerId: string;
  originPort: string;
  destinationPort: string;
  departureDate: string; // ISO date string
  status?: ShipmentStatus;
  eta: string; // ISO date string, computed by frontend
  lastKnownLat?: number;
  lastKnownLng?: number;
}

export interface UpdateShipmentRequest {
  status?: ShipmentStatus;
  eta?: string;
  lastKnownLat?: number;
  lastKnownLng?: number;
}

export interface ShipmentSearchQuery {
  search?: string;
  status?: ShipmentStatus;
  limit?: number;
  offset?: number;
}

File: client/Models/Shipment.cs
using System.ComponentModel.DataAnnotations;

namespace client.Models;

public enum ShipmentStatus
{
    InTransit,
    ArrivedAtPort,
    Delayed,
    Delivered
}

public class Shipment
{
    public int Id { get; set; }
    
    [Required]
    [StringLength(50)]
    public string ContainerId { get; set; } = string.Empty;
    
    [Required]
    [StringLength(100)]
    public string OriginPort { get; set; } = string.Empty;
    
    [Required]
    [StringLength(100)]
    public string DestinationPort { get; set; } = string.Empty;
    
    [Required]
    public DateTime DepartureDate { get; set; }
    
    public string Status { get; set; } = "In Transit";
    
    [Required]
    public DateTime Eta { get; set; }
    
    public double? LastKnownLat { get; set; }
    
    public double? LastKnownLng { get; set; }
    
    public DateTime CreatedAt { get; set; }
    
    public DateTime UpdatedAt { get; set; }
}

// API Response Models
public class ApiResponse<T>
{
    public T Data { get; set; } = default!;
    public Pagination? Pagination { get; set; }
}

public class Pagination
{
    public int Total { get; set; }
    public int Limit { get; set; }
    public int Offset { get; set; }
    public bool HasMore { get; set; }
}

public class CreateShipmentRequest
{
    [Required]
    [StringLength(50)]
    public string ContainerId { get; set; } = string.Empty;
    
    [Required]
    [StringLength(100)]
    public string OriginPort { get; set; } = string.Empty;
    
    [Required]
    [StringLength(100)]
    public string DestinationPort { get; set; } = string.Empty;
    
    [Required]
    public DateTime DepartureDate { get; set; }
    
    public string Status { get; set; } = "In Transit";
    
    [Required]
    public DateTime Eta { get; set; }
    
    public double? LastKnownLat { get; set; }
    
    public double? LastKnownLng { get; set; }
}

public class UpdateShipmentRequest
{
    public string? Status { get; set; }
    
    public DateTime? Eta { get; set; }
    
    public double? LastKnownLat { get; set; }
    
    public double? LastKnownLng { get; set; }
}
